promql/engine.go

```diff
@@ -3419,6 +3478,65 @@ func makeInt64Pointer(val int64) *int64 {
 	return valp
 }
 
+// Add RatioSampler interface to allow unit-testing (previously: Randomizer).
+type RatioSampler interface {
+	// Return this sample "offset" between [0.0, 1.0]
+	sampleOffset(ts int64, sample *Sample) float64
+	AddRatioSample(r float64, sample *Sample) bool
+}
+
+// Use Hash(labels.String()) / maxUint64 as a "deterministic"
+// value in [0.0, 1.0].
+type HashRatioSampler struct{}
+
+var ratiosampler RatioSampler = NewHashRatioSampler()
+
+func NewHashRatioSampler() *HashRatioSampler {
+	return &HashRatioSampler{}
+}
+
+func (s *HashRatioSampler) sampleOffset(ts int64, sample *Sample) float64 {
+	const (
+		float64MaxUint64 = float64(math.MaxUint64)
+	)
+	return float64(sample.Metric.Hash()) / float64MaxUint64
+}
+
+func (s *HashRatioSampler) AddRatioSample(ratioLimit float64, sample *Sample) bool {
+	sampleOffset := s.sampleOffset(sample.T, sample)
+	switch {
+	case ratioLimit >= 0:
+		// If ratioLimit >= 0: add sample if ratiosampler.sampleOffset() is lesser than ratioLimit
+		//
+		// 0.0        ratioLimit                1.0
+		//  [---------|--------------------------]
+		//  [#########...........................]
+		//
+		// e.g.:
+		//   ratiosampler.sampleOffset()==0.3 && ratioLimit==0.4
+		//     0.3 < 0.4 ? --> add sample
+		//
+		if sampleOffset < ratioLimit {
```

Comment:
I realize this wasn't a `return` in past iterations, but with the current code structure it would probably be more readable to just return the result of the boolean comparison directly in both cases and not require the `return false` at the end at all.

Reply:
Oh good catch, fixed at 1188b10c.